pot/pot.go

```diff
@@ -0,0 +1,793 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package pot
+
+import (
+	"fmt"
+	"sync"
+)
+
+const (
+	// keylen = 4
+	keylen    = 256
+	maxkeylen = 256
+)
+
+// Pot is the root node type, allows locked non-applicative manipulation
+type Pot struct {
+	lock sync.RWMutex
+	*pot
+}
+
+// pot is the node type (same for root, branching node and leaf)
+type pot struct {
+	pin  PotVal
+	bins []*pot
+	size int
+	po   int
+}
+
+// PotVal is the interface the generic container item should implement
+type PotVal interface {
+	PO(PotVal, int) (po int, eq bool)
+	String() string
+}
+
+// Pot constructor. Requires  value of type PotVal to pin
+// and po to point to a span in the PotVal key
+// The pinned item counts towards the size
+func NewPot(v PotVal, po int) *Pot {
+	var size int
+	if v != nil {
+		size++
+	}
+	return &Pot{
+		pot: &pot{
+			pin:  v,
+			po:   po,
+			size: size,
+		},
+	}
+}
+
+// Pin() returns the pinned element (key) of the Pot
```

Comment:
Please omit `()` and parameter names from the beginning of documentation comments.